src/process/io/s3/S3Sink.h

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_S3SINK_H
+#define FASTQINDEX_S3SINK_H
+
+#include "common/StringHelper.h"
+#include "process/io/PathSink.h"
+#include "process/io/Sink.h"
+#include "S3Config.h"
+#include "FQIS3Client.h"
+#include <cstdio>
+#include <fcntl.h>
+#include <memory>
+
+#include <aws/core/auth/AWSAuthSigner.h>
+#include <aws/core/auth/AWSCredentialsProvider.h>
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/Aws.h>
+#include <aws/s3/model/ListObjectsRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/s3/S3Client.h>
+
+using namespace Aws::Utils;
+using namespace Aws::S3;
+using namespace Aws::S3::Model;
+
+class S3Sink : public Sink {
+private:
+
+    bool _isOpen{false};
+
+    bool objectAlreadyExists{false};
+
+    PathSink *tempFile{nullptr};
+
+    FQIS3Client fqiS3Client;
+
+public:
+
+    static shared_ptr<S3Sink> from(const string &file, bool forceOverwrite, const S3ServiceOptions &s3ServiceOptions) {
+        return make_shared<S3Sink>(file, forceOverwrite, s3ServiceOptions);
+    }
+
+    S3Sink(const string &s3Path, bool forceOverwrite, const S3ServiceOptions &s3ServiceOptions) :
+            fqiS3Client(s3Path, s3ServiceOptions),
+            Sink(forceOverwrite) {
+    }
+
+    ~S3Sink() override {
+        close();
+    }
+
+    bool openWithWriteLock() override {
+        return open();
+    }
+
+    bool fulfillsPremises() override {
+        return true;
+    }
+
+
+    bool open() override {
+        if (isOpen())
+            return true;
+
+        if (!fqiS3Client.isValid()) {
+            return false;
+        }
+
+        // Assign these values before running the program
+        bool ok = true;
+
+        auto result = fqiS3Client.checkObjectExistence();
+        if (!result.requestWasSuccessful)
+            return false;
+
+        objectAlreadyExists = result.result;
+
+        if (objectAlreadyExists && !this->forceOverwrite) {
+            addErrorMessage("The file '", this->fqiS3Client.getS3Path(),
+                            "' already exists. ",
+                            "Use -w to force the application to overwrite the file.");
```

Comment:
Hm. I really like to see helpful error messages. Hm. I really like to have CLI information outside of core code. Hm ...